perllib/FixMyStreet/App/Controller/Admin/ManifestTheme.pm

```diff
@@ -0,0 +1,99 @@
+
+use FixMyStreet::App::Form::ManifestTheme;
+
+sub auto :Private {
+    my ($self, $c) = @_;
+
+    if ( $c->cobrand->moniker eq 'fixmystreet' ) {
+        $c->stash(rs => $c->model('DB::ManifestTheme')->search_rs({}), show_all => 1);
+    } else {
+        $c->stash(rs => $c->model('DB::ManifestTheme')->search_rs({ cobrand => $c->cobrand->moniker }));
+    }
+}
+
+sub index :Path :Args(0) {
+    my ( $self, $c ) = @_;
+
+    unless ( $c->stash->{show_all} ) {
+        if ( $c->stash->{rs}->count ) {
+            $c->res->redirect($c->uri_for($self->action_for('edit'), [ $c->stash->{rs}->first->cobrand ]));
+        } else {
+            $c->res->redirect($c->uri_for($self->action_for('create')));
+        }
+        $c->detach;
+    }
+}
+
+sub item :PathPart('admin/manifesttheme') :Chained :CaptureArgs(1) {
+    my ($self, $c, $cobrand) = @_;
+
+    my $obj = $c->stash->{rs}->find({ cobrand =>  $cobrand })
+        or $c->detach('/page_error_404_not_found', []);
+    $c->stash(obj => $obj);
+}
+
+sub edit :PathPart('') :Chained('item') :Args(0) {
+    my ($self, $c) = @_;
+
+    my $form = $self->form($c, $c->stash->{obj});
+
+    # We need to do this after form processing, in case a form POST has deleted
+    # an icon.
+    $c->forward('/offline/_stash_manifest_theme', [ $c->stash->{obj}->cobrand, 1 ]);
+
+    return $form;
+}
+
+
+sub create :Local :Args(0) {
+    my ($self, $c) = @_;
+
+    unless ( $c->stash->{show_all} || $c->stash->{rs}->count == 0) {
+        $c->res->redirect($c->uri_for($self->action_for('edit'), [ $c->stash->{rs}->first->cobrand ]));
+        $c->detach;
+    }
+
+    my $theme = $c->stash->{rs}->new_result({});
+    return $self->form($c, $theme);
+}
+
+sub form {
+    my ($self, $c, $theme) = @_;
+
+    if ($c->get_param('delete_theme')) {
+        $c->forward('_delete_all_manifest_icons');
+        $c->forward('/offline/_clear_manifest_theme_cache', [ $theme->cobrand ]);
+        $theme->delete;
+        $c->forward('/admin/log_edit', [ $theme->id, 'manifesttheme', 'delete' ]);
+        $c->response->redirect($c->uri_for($self->action_for('index')));
+        $c->detach;
+    }
+
+    my $action = $theme->in_storage ? 'edit' : 'add';
+    my $form = FixMyStreet::App::Form::ManifestTheme->new( cobrand => $c->cobrand->moniker );
+    $c->stash(template => 'admin/manifesttheme/form.html', form => $form);
+    my $params = $c->req->params;
+    $params->{icon} = $c->req->upload('icon') if $params->{icon};
+    $form->process(item => $theme, params => $params);
+    return unless $form->validated;
+
+    $c->forward('/admin/log_edit', [ $theme->id, 'manifesttheme', $action ]);
+    $c->forward('/offline/_clear_manifest_theme_cache', [ $theme->cobrand ]);
+    $c->response->redirect($c->uri_for($self->action_for('index')));
+}
+
+sub _delete_all_manifest_icons :Private {
+    my ($self, $c) = @_;
+
+    $c->forward('/offline/_stash_manifest_theme', [ $c->stash->{obj}->cobrand, 1 ]);
```

Comment:
I guess if it was done on every request, you wouldn't need to call this here? But no harm in it.

Reply:
Actually no, because the call in `Root::auto` will include the default theme & icons if there isn't one present in the DB. This call will ignore any default icons on disk and memcache.

Reply:
Hmm, so does that mean those admin pages lose the icons/theme-colour in their header? Or if I'm editing a council cobrand's theme on .com, that page is themed with the cobrand? Just feels like something could get confused - it feels like it should be able to have the stashed theme for the page, and a separate theme for being edited? Or have I misunderstood.

Reply:
You're right - I hadn't thought about the current theme of the page and the theme you're editing as two separate things. I've tidied things up and made it clearer in 6074fdb3362993853b1292f069170d52bb83b36b.